### PR TITLE
Make route controller care about faults

### LINKED ISSUE #521: Make route controller care about faults
## User stories
As the fault injector I want the RouteController to recalculate all or all relevant routes in order for trains not to get stuck or drive over blocked tracks.

## Dependencies
Depends on: 
 - #438 

## (Wiki) References
- [Architecture](https://github.com/BP2022-AP1/bp2022-ap1/wiki/Interlocking-Component#architecture) of the Interlocking Component

## Acceptance Criteria
- [x] The `IInterlockingDisruptor` calls the `RouteController` to recalculate all routes
    - [x] `insert_track_blocked`
    - [x] `insert_track_unblocked`
    - [x] `insert_train_priority_changed`
- [x] `RouteController` implemts the method `recalculate_all_routes`
    - [x] `routes_to_be_set` and `routes_to_be_reserved` is set to an empty list
    - [x] it gets all Trains currently in the Simulation from the `SimulationObjectUpdatingComponent`
    - [x] for each Train `_free_fahrstrasse` is called
    - [x] then for each Train `set_fahrstrasse` is called
- [x] use or delete `check_all_fahrstrassen_for_failures` method from `RouteController`

### PR DESCRIPTION
**Pull request by instant-sky**
Fixes #521 

Description (what might a PO or reviewer want to know?)

## PR checklist

- [ ] Acceptance criteria fulfilled
- [ ] Additional features are tested
- [ ] Docs (code, wiki & diagrams) updated
- [ ] Breaking changes anounced
- [ ] Added classes that inherit from `BaseModel` to `src.constants.tables`
- [ ] Dev-branch has been merged into local branch to resolve conflicts
- [ ] Tests and linter have passed AFTER local merge
- [ ] Another dev reviewed and approved

### CODE REVIEW

File: src/interlocking_component/route_controller.py [hunk 1]
```diff
@@ -55,7 +61,7 @@ def insert_platform_unblocked(self, platform: Platform):
         :param platform: the unblocked platform
         :type platform: Platform
         """
-        raise NotImplementedError()
+        self.route_controller.recalculate_all_routes()
 
     def insert_track_speed_limit_changed(self, track: Track):
```
Comment (Lietze) [inline src/interlocking_component/route_controller.py]: I am wondering, weather these two methods should (for now) just be deleted. Speed of trains or the tracks is not (yet) taken into account.

File: src/interlocking_component/route_controller.py [hunk 2]
```diff
@@ -330,6 +338,14 @@ def _get_interlocking_routes_for_edge(self, edge: Edge) -> List[Route]:
                 routes.append(route_candidate)
         return routes
 
-    def check_all_fahrstrassen_for_failures(self):
-        """This method checks for all trains, if their fahrstrassen and routes are still valid."""
-        raise NotImplementedError()
+    # def check_all_fahrstrassen_for_failures(self):
+    #    """This method checks for all trains, if their fahrstrassen and routes are still valid."""
+    #    raise NotImplementedError()
```
Comment (Lietze) [inline src/interlocking_component/route_controller.py]: Just remove this.
```suggestion
```

File: src/interlocking_component/route_controller.py [hunk 3]
```diff
@@ -330,6 +338,14 @@ def _get_interlocking_routes_for_edge(self, edge: Edge) -> List[Route]:
...
+
+    def recalculate_all_routes(self):
+        """Recalculates the route for every train in the simulation"""
+        self.routes_to_be_set = []
```
Comment (Lietze) [inline src/interlocking_component/route_controller.py:348]: If this get's merged after #438 yoe need to remember to also set routes_to_be_reserved to an empty List.

File: tests/fault_injector/conftest.py [hunk 4]
```diff
@@ -49,9 +49,16 @@ def event_bus(run):
...
 
+class MockRouteController:
+    method_calls: int = 0
+
+    def recalculate_all_routes(self):
+        self.method_calls += 1
+
+
 @pytest.fixture
 def interlocking():
```
Comment (Lietze) [inline tests/fault_injector/conftest.py]: Here again, I do not realy like the naming. But it is not that important.

File: tests/fault_injector/test_platform_blocked_fault.py [hunk 5]
```diff
@@ -61,8 +61,8 @@ def test_inject_platform_blocked_fault(
         combine_platform_and_wrapper,
     ):
         assert not platform.blocked
-        with pytest.raises(NotImplementedError):
-            platform_blocked_fault.inject_fault(tick)
+        platform_blocked_fault.inject_fault(tick)
+        assert platform_blocked_fault.interlocking.route_controller.method_calls > 0
```
Comment (Lietze) [inline tests/fault_injector/test_platform_blocked_fault.py]: Why not exactly 1?
```suggestion
        assert platform_blocked_fault.interlocking.route_controller.method_calls == 1
```

File: tests/fault_injector/test_platform_blocked_fault.py [hunk 6]
```diff
@@ -74,9 +74,9 @@ def test_resolve_platform_blocked_fault(
...
     ):
-        with pytest.raises(NotImplementedError):
-            platform_blocked_fault.inject_fault(tick)
+        platform_blocked_fault.inject_fault(tick)
         assert platform.blocked
-        with pytest.raises(NotImplementedError):
-            platform_blocked_fault.resolve_fault(tick)
+        assert platform_blocked_fault.interlocking.route_controller.method_calls > 0
+        platform_blocked_fault.resolve_fault(tick)
+        assert platform_blocked_fault.interlocking.route_controller.method_calls > 1
```
Comment (Lietze) [inline tests/fault_injector/test_platform_blocked_fault.py]: Again, why not use exact numbers?
```suggestion
        assert platform_blocked_fault.interlocking.route_controller.method_calls == 1
        platform_blocked_fault.resolve_fault(tick)
        assert platform_blocked_fault.interlocking.route_controller.method_calls == 2
```

File: tests/fault_injector/test_track_blocked_fault.py [hunk 7]
```diff
@@ -57,8 +57,8 @@ def test_inject_track_blocked_fault(
         combine_track_and_wrapper,
     ):
         assert not track.blocked
-        with pytest.raises(NotImplementedError):
-            track_blocked_fault.inject_fault(tick)
+        track_blocked_fault.inject_fault(tick)
+        assert track_blocked_fault.interlocking.route_controller.method_calls > 0
```
Comment (Lietze) [inline tests/fault_injector/test_track_blocked_fault.py]:
```suggestion
        assert track_blocked_fault.interlocking.route_controller.method_calls == 1
```

File: tests/fault_injector/test_track_blocked_fault.py [hunk 8]
```diff
@@ -70,9 +70,9 @@ def test_resolve_track_blocked_fault(
...
-        with pytest.raises(NotImplementedError):
-            track_blocked_fault.inject_fault(tick)
+        track_blocked_fault.inject_fault(tick)
         assert track.blocked
-        with pytest.raises(NotImplementedError):
-            track_blocked_fault.resolve_fault(tick)
+        assert track_blocked_fault.interlocking.route_controller.method_calls > 0
+        track_blocked_fault.resolve_fault(tick)
         assert not track.blocked
+        assert track_blocked_fault.interlocking.route_controller.method_calls > 1
```
Comment (Lietze) [inline tests/fault_injector/test_track_blocked_fault.py]:
```suggestion
        assert track_blocked_fault.interlocking.route_controller.method_calls == 1
        track_blocked_fault.resolve_fault(tick)
        assert not track.blocked
        assert track_blocked_fault.interlocking.route_controller.method_calls == 2
```

File: tests/fault_injector/test_track_speed_limit_fault.py [hunk 9]
```diff
@@ -70,9 +70,9 @@ def test_inject_track_speed_limit_fault(
     ):
         track.max_speed = 100
         assert track.max_speed == 100
-        with pytest.raises(NotImplementedError):
-            track_speed_limit_fault.inject_fault(tick=tick)
+        track_speed_limit_fault.inject_fault(tick=tick)
         assert track.max_speed == track_speed_limit_fault.configuration.new_speed_limit
+        assert track_speed_limit_fault.interlocking.route_controller.method_calls > 0
```
Comment (Lietze) [inline tests/fault_injector/test_track_speed_limit_fault.py]:
```suggestion
        assert track_speed_limit_fault.interlocking.route_controller.method_calls == 1

```

File: tests/fault_injector/test_track_speed_limit_fault.py [hunk 10]
```diff
@@ -86,8 +86,8 @@ def test_resolve_track_speed_limit_fault(
...
         track.max_speed = 100
-        with pytest.raises(NotImplementedError):
-            track_speed_limit_fault.inject_fault(tick=tick)
-        with pytest.raises(NotImplementedError):
-            track_speed_limit_fault.resolve_fault(tick=tick)
+        track_speed_limit_fault.inject_fault(tick=tick)
+        assert track_speed_limit_fault.interlocking.route_controller.method_calls > 0
+        track_speed_limit_fault.resolve_fault(tick=tick)
         assert track.max_speed == track_speed_limit_fault.old_speed_limit
+        assert track_speed_limit_fault.interlocking.route_controller.method_calls > 1
```
Comment (Lietze) [inline tests/fault_injector/test_track_speed_limit_fault.py]:
```suggestion
        assert track_speed_limit_fault.interlocking.route_controller.method_calls == 1
        track_speed_limit_fault.resolve_fault(tick=tick)
        assert track.max_speed == track_speed_limit_fault.old_speed_limit
        assert track_speed_limit_fault.interlocking.route_controller.method_calls == 2

```

File: tests/fault_injector/test_train_prio_fault.py [hunk 11]
```diff
@@ -58,9 +58,9 @@ def test_inject_train_prio_fault(
         combine_train_and_wrapper,
     ):
         train.train_type.priority = 1
-        with pytest.raises(NotImplementedError):
-            train_prio_fault.inject_fault(tick)
+        train_prio_fault.inject_fault(tick)
         assert train.train_type.priority == train_prio_fault.configuration.new_prio
+        assert train_prio_fault.interlocking.route_controller.method_calls > 0
```
Comment (Lietze) [inline tests/fault_injector/test_train_prio_fault.py]:
```suggestion
        assert train_prio_fault.interlocking.route_controller.method_calls == 1
```

File: tests/fault_injector/test_train_prio_fault.py [hunk 12]
```diff
@@ -72,12 +72,12 @@ def test_resolve_train_prio_fault(
...
-        with pytest.raises(NotImplementedError):
-            train_prio_fault.inject_fault(tick)
+        train_prio_fault.inject_fault(tick)
         assert train.train_type.priority == train_prio_fault.configuration.new_prio
-        with pytest.raises(NotImplementedError):
-            train_prio_fault.resolve_fault(tick)
+        assert train_prio_fault.interlocking.route_controller.method_calls > 0
+        train_prio_fault.resolve_fault(tick)
         assert train.train_type.priority == train_prio_fault.old_prio == 5
+        assert train_prio_fault.interlocking.route_controller.method_calls > 1
```
Comment (Lietze) [inline tests/fault_injector/test_train_prio_fault.py]:
```suggestion
        assert train_prio_fault.interlocking.route_controller.method_calls == 1
        train_prio_fault.resolve_fault(tick)
        assert train.train_type.priority == train_prio_fault.old_prio == 5
        assert train_prio_fault.interlocking.route_controller.method_calls == 2
```

File: tests/fault_injector/test_train_speed_fault.py [hunk 13]
```diff
@@ -72,9 +72,9 @@ def test_inject_train_speed_fault(
         # pylint: enable=unused-argument
     ):
         train.train_type.max_speed = 100
-        with pytest.raises(NotImplementedError):
-            train_speed_fault.inject_fault(tick=tick)
+        train_speed_fault.inject_fault(tick=tick)
         assert train.train_type.max_speed == train_speed_fault.configuration.new_speed
+        assert train_speed_fault.interlocking.route_controller.method_calls > 0
```
Comment (Lietze) [inline tests/fault_injector/test_train_speed_fault.py]:
```suggestion
        assert train_speed_fault.interlocking.route_controller.method_calls == 1
```

File: tests/fault_injector/test_train_speed_fault.py [hunk 14]
```diff
@@ -89,12 +89,12 @@ def test_resolve_train_speed_fault(
...
-        with pytest.raises(NotImplementedError):
-            train_speed_fault.inject_fault(tick=tick)
+        train_speed_fault.inject_fault(tick=tick)
         assert train.train_type.max_speed == train_speed_fault.configuration.new_speed
-        with pytest.raises(NotImplementedError):
-            train_speed_fault.resolve_fault(tick=tick)
+        assert train_speed_fault.interlocking.route_controller.method_calls > 0
+        train_speed_fault.resolve_fault(tick=tick)
         assert train.train_type.max_speed == train_speed_fault.old_speed == 50
+        assert train_speed_fault.interlocking.route_controller.method_calls > 1
```
Comment (Lietze) [inline tests/fault_injector/test_train_speed_fault.py]:
```suggestion
        assert train_speed_fault.interlocking.route_controller.method_calls == 1
        train_speed_fault.resolve_fault(tick=tick)
        assert train.train_type.max_speed == train_speed_fault.old_speed == 50
        assert train_speed_fault.interlocking.route_controller.method_calls == 2
```

**Review by Lietze — CHANGES_REQUESTED** [review]
Great job! Thank you for implementing this.
Do you think, that in fault.py the `interlocking` should be renamed to `interlocking_disruptor`? That may be better, because the interlocking is a resource from outside.

File: src/interlocking_component/route_controller.py (hunk 1 above)
Comment (instant-sky) [reply src/interlocking_component/route_controller.py]: What do you think about commenting them out and adding a comment why it is like that at the moment?

File: src/interlocking_component/route_controller.py (hunk 3 above)
Comment (instant-sky) [reply src/interlocking_component/route_controller.py:348]: I added a comment which can be commented in in this case

File: tests/fault_injector/conftest.py (hunk 4 above)
Comment (instant-sky) [reply tests/fault_injector/conftest.py]: I changed it to `interlocking_disruptor`

File: src/interlocking_component/route_controller.py (hunk 1 above)
Comment (Lietze) [reply src/interlocking_component/route_controller.py]: Yes, that sounds even better

**Review by Lietze — APPROVED** [review]
Looks good, thank you!

**Comment by antonykamp** [comment]
Great teamwork guys 🙌

**Pull request merged**